### PR TITLE
Add `aspire.utils.coor_trans.crop_2d` and move `coor_trans` objects to second-level API access

### LINKED ISSUE #140: Downsample bug
Seems like downsample in `src/aspire/image/__init__.py` doesn't work as the matlab code.
I can put replace the crop and downsample functions in `src/aspire/utils/preprocess.py` with my code [here](https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/itay_merge/src/aspire/preprocess/preprocessor.py) line 88 and 104.

The implementation is a lot simpler, but for now it doesn't support up sampling (only crop needs to be revised to do so).

### PR DESCRIPTION
**Pull request by chris-langfield**
Following the discussion in issue #140 we decided to replace the `Image.downsample` method with a port of the MATLAB algorithm, specifically `cryo_downsample`. This algorithm uses a square crop in Fourier space. This PR adds a partial port of `cryo_crop` from the MATLAB code. 

It also addresses #554 by moving all functions in `aspire.utils.coor_trans` to be accessed directly from `aspire.utils`. 

There is a bug in the MATLAB code for padding from size `n` to `n+1` when `n` is even. Below is the exact logic of the matlab `cryo_crop` logic for 2D (in Python):

```python
def crop_2d(mat, size, fill_value=0):
    mat_x, mat_y = mat.shape
    start_x = math.floor(mat_x / 2) - math.floor(size / 2)
    start_y = math.floor(mat_y / 2) - math.floor(size / 2)

    # cropping
    if start_x >= 0 and start_y >= 0:
        return mat[start_x : start_x + size, start_y : start_y + size]
    # padding
    elif start_x < 0 and start_y < 0:
        to_return = fill_value * np.ones((size, size))
        to_return[start_x : start_x + size, start_y : start_y + size] = mat
        return to_return
```

The check of `start_x` and `start_y` is supposed to distinguish between cropping and padding. However, an edge case occurs when `n` is even and we are padding to size `n+1`. In this case, `start_x = start_y = 0`, so the *cropping* code is run. However, because `size > mat_x` and `size > mat_y`, we get an index out of bounds error. The implementation in this PR fixes this by slightly modifying the check of `start_x` and `start_y` and adding a comparison of the size to `mat_x` and `mat_y`. 

Reproduce in MATLAB:

```
inm = diag([1 2 3 4 5 6 7 8 9 10])
out = cryo_crop(inm, 11)

Index in position 1 exceeds array bounds (must not exceed 10).

Error in cryo_crop (line 55)
            mc=m(nsx+1:nsx+n(1),nsy+1:nsy+n(2));
```

### CODE REVIEW

**Comment by chris-langfield** [comment]
I realized that the old `downsample` and `crop_pad` methods exist in `aspire/image/preprocess.py` still. However, they are only used in `test_preprocess.py`. 

`crop_pad` in particular has the behavior of `crop_2d` as a special case, but without the fix described above. 

I'm of the opinion that for this targeted change (changing the downsampling implementation), it's better to use this smaller method than try to use/modify `aspire.image.preprocess.crop_pad`. We can figure out breaking up `crop_pad` into more specialized `utils` methods in future work.

File: src/aspire/utils/coor_trans.py [hunk 1]
```diff
@@ -1,7 +1,9 @@
-"""
+"""OA
```
Comment (j-c-c) [inline src/aspire/utils/coor_trans.py]: Looks like some mistaken characters here:  `OA`

File: src/aspire/utils/coor_trans.py [hunk 2]
```diff
@@ -302,3 +304,31 @@ def common_line_from_rots(r1, r2, ell):
...
+    if start_x >= 0 and start_y >= 0 and size < min(mat_x, mat_y):
+        return mat[start_x : start_x + size, start_y : start_y + size]
+    # padding
+    elif start_x <= 0 and start_y <= 0:
+        # ensure that we return in the same dtype as the input
+        to_return = fill_value * np.ones((size, size), dtype=mat.dtype)
+        to_return[-start_x : mat_x - start_x, -start_y : mat_y - start_y] = mat
+        return to_return
+    else:
+        raise ValueError("Cannot crop and pad an image at the same time.")
```
Comment (j-c-c) [inline src/aspire/utils/coor_trans.py]: This looks good! One possible alteration:

I might be missing some corner case here, but I think these conditionals can be simplified to:
`if size <= min(mat_x, mat_y):`
`elif size > max(mat_x, mat_y):`
`else:`

File: tests/test_coor_trans.py [hunk 3]
```diff
@@ -77,3 +78,87 @@ def testRegisterRots(self):
         q_mat_est, flag_est = register_rotations(rots_ref, regrots_ref)
 
         self.assertTrue(np.allclose(flag_est, flag) and np.allclose(q_mat_est, q_mat))
+
+    def testSquareCrop2D(self):
```
Comment (j-c-c) [inline tests/test_coor_trans.py:82]: Maybe test a rectangular M x N matrix, M != N, with `size` in between M and N. That way you'll get coverage for the `else:` condition in `crop_2d` (line 334).

**Review by j-c-c — CHANGES_REQUESTED** [review]
Looks great! A couple small things. Double check my logic on the "conditionals" suggestion.

File: src/aspire/utils/coor_trans.py (hunk 2 above)
Comment (j-c-c) [reply src/aspire/utils/coor_trans.py]: Oops, maybe that `elif` should be:

`elif size >= max(mat_x, mat_y):`

**Review by j-c-c — DISMISSED** [review]
Did a brief sweep of the changes. Looks good to me.

File: src/aspire/utils/coor_trans.py [hunk 4]
```diff
@@ -302,3 +304,29 @@ def common_line_from_rots(r1, r2, ell):
     ell_ji = int(np.mod(np.round(ell_ji), ell))
 
     return ell_ij, ell_ji
+
+
+def crop_2d(mat, size, fill_value=0):
```
Comment (janden) [inline src/aspire/utils/coor_trans.py]: I'm wondering if there's a better name here, since we're doing both cropping and padding. Maybe `rebox`, but that seems ugly?

An alternative is to define `pad_2d = crop_2d` further down.

File: src/aspire/utils/coor_trans.py [hunk 5]
```diff
@@ -302,3 +304,29 @@ def common_line_from_rots(r1, r2, ell):
...
+    """
+    :param mat: A 2-dimensional numpy array
+    :param size: Integer size of cropped/padded output
+    :return: A numpy array of shape (size, size)
+    """
+
+    mat_x, mat_y = mat.shape
```
Comment (janden) [inline src/aspire/utils/coor_trans.py]: It doesn't matter much, but it would be better to stick YX since that's what we're doing elsewhere, so we'd have `mat_y, mat_x = mat.shape`, etc.

File: src/aspire/utils/coor_trans.py [hunk 6]
```diff
@@ -302,3 +304,29 @@ def common_line_from_rots(r1, r2, ell):
     ell_ji = int(np.mod(np.round(ell_ji), ell))
 
     return ell_ij, ell_ji
+
+
+def crop_2d(mat, size, fill_value=0):
+    """
+    :param mat: A 2-dimensional numpy array
```
Comment (janden) [inline src/aspire/utils/coor_trans.py]: Maybe `im` instead of `mat` since this is not a matrix.

Comment (janden) [inline src/aspire/utils/coor_trans.py]: Also, should we restrict the input to square since we are restricted to square outputs?

File: src/aspire/utils/coor_trans.py [hunk 7]
```diff
@@ -302,3 +304,29 @@ def common_line_from_rots(r1, r2, ell):
...
+    start_y = math.floor(mat_y / 2) - math.floor(size / 2)
+
+    # cropping
+    if size <= min(mat_x, mat_y):
+        return mat[start_x : start_x + size, start_y : start_y + size]
+    # padding
+    elif size >= max(mat_x, mat_y):
+        # ensure that we return in the same dtype as the input
+        to_return = fill_value * np.ones((size, size), dtype=mat.dtype)
+        to_return[-start_x : mat_x - start_x, -start_y : mat_y - start_y] = mat
```
Comment (janden) [inline src/aspire/utils/coor_trans.py]: I'm confused. Why the negative indices?

**Review by janden — CHANGES_REQUESTED** [review]
Just some quick comments. Haven't looked at the tests yet.

File: src/aspire/utils/coor_trans.py (hunk 6 above)
Comment (chris-langfield) [reply src/aspire/utils/coor_trans.py]: agreed and done

File: src/aspire/utils/coor_trans.py (hunk 4 above)
Comment (chris-langfield) [reply src/aspire/utils/coor_trans.py]: I am going with `crop_pad_2d`. I think the problem with rebox, resize, etc. is that there's some semantic overlap with downsampling. Having the word "crop" emphasizes that the resolution is not changed

File: src/aspire/utils/coor_trans.py (hunk 7 above)
Comment (chris-langfield) [reply src/aspire/utils/coor_trans.py]: This is from the original implementation. `start_x` and `start_y` are actually themselves negative when cropping because `size` is larger than `mat_x` and `mat_y`.

```python
# shift terms                                                                                                          
    start_x = math.floor(im_x / 2) - math.floor(size / 2)
    start_y = math.floor(im_y / 2) - math.floor(size / 2)
```

Comment (chris-langfield) [reply src/aspire/utils/coor_trans.py]: It basically calculates the starting point of where we put the original image, relative to the new (larger) box / frame around it.

Comment (janden) [reply src/aspire/utils/coor_trans.py]: Huh. May be worth a clarifying comment.

**Review by janden — CHANGES_REQUESTED** [review]
Looks good. Just one small thing.

**Review by janden — APPROVED** [review]
Thanks!
